mlrun/auth/providers.py

```
@@ -345,17 +360,34 @@ def _parse_response(self, response_data: dict, raise_on_error=False):
         access_token = spec.get("accessToken")
 
         if not access_token:
-            mlrun.utils.helpers.raise_or_log_error(
-                "Access token is missing in the response from the token endpoint",
-                raise_on_error,
+            raise mlrun.errors.MLRunRuntimeError(
+                "Access token is missing in the response from the token endpoint"
             )
-            return
 
         self._token = access_token
+
         self._token_total_lifetime, self._token_expiry_time = (
             self._get_token_lifetime_and_expiry(access_token)
         )
 
+    def _post_fetch_hook(self):
+        # After fetching a new token, sync the local token file with the backend to ensure the latest token is stored
+        mlrun.utils.helpers.run_with_retry(
+            retry_count=self._max_retries,
+            func=mlrun.secrets.sync_secret_tokens,
+        )
```

sync tokens should be called last. If self._token is None, it will raise an error, meaning you end up syncing something unnecessary.

---

@moranbental not sure I agree here. In provider we work with a single `default` token, but this function syncs all of the tokens. So I think what user expects is that all the tokens are synced no matter what's wrong with the default one.

---

ok, so it’s worth adding a comment on that here

---

@moranbental the question is do we want to raise an exception if sync didn't succeed or not?

---

@rokatyy so I’m not sure we want to raise an exception here, since it could be spammy for the user.
@theSaarco / @quaark correct me if I'm wrong.

---

@moranbental also not sure, because how else user will know that something is wrong with the tokens sync? is there supposed to be any other flow where it will be synced?

---

user can call `mlrun.sync_secret_tokens` directly.

---

Spoke with @moranbental offline. I think there's been some confusion here. The sync tokens should not happen on every api call. Only when importing MLRun. Or if the user ran the method specifically

---

Make sense, removing it from Token provider